File: backend/app/core/seed.py
import asyncio

from sqlalchemy import select

from app.core.config import settings
from app.core.database import AsyncSessionLocal
from app.models.entities import Area, Building, City, Country, Street, User
from app.models.enums import UserRole
from app.services.text import normalize_name


async def seed() -> None:
    async with AsyncSessionLocal() as db:
        country = (await db.execute(select(Country).where(Country.code == "PT"))).scalar_one_or_none()
        if not country:
            country = Country(code="PT", name_en="Portugal", name_pt="Portugal")
            db.add(country)
            await db.flush()

        for city_name in ["Lisboa", "Porto"]:
            city = (
                await db.execute(
                    select(City).where(City.country_id == country.id, City.normalized_name == normalize_name(city_name))
                )
            ).scalar_one_or_none()
            if not city:
                city = City(country_id=country.id, name=city_name, normalized_name=normalize_name(city_name))
                db.add(city)
                await db.flush()

                area_name = "Avenidas Novas" if city_name == "Lisboa" else "Cedofeita"
                street_name = "Avenida da República" if city_name == "Lisboa" else "Rua de Cedofeita"

                area = Area(city_id=city.id, name=area_name, normalized_name=normalize_name(area_name))
                db.add(area)
                await db.flush()

                street = Street(area_id=area.id, name=street_name, normalized_name=normalize_name(street_name))
                db.add(street)
                await db.flush()

                building = Building(
                    street_id=street.id,
                    segment_id=None,
                    street_number=100 if city_name == "Lisboa" else 250,
                    building_name=f"LivedHere {city_name} Demo",
                    lat=38.736946 if city_name == "Lisboa" else 41.157944,
                    lng=-9.142685 if city_name == "Lisboa" else -8.629105,
                )
                db.add(building)

        admin_emails = settings.admin_emails

        current_admins = (await db.execute(select(User).where(User.role == UserRole.ADMIN))).scalars().all()
        for admin_user in current_admins:
            if admin_user.email.strip().lower() not in admin_emails:
                admin_user.role = UserRole.USER

        for admin_email in admin_emails:
            admin = (await db.execute(select(User).where(User.email == admin_email))).scalar_one_or_none()
            if not admin:
                db.add(User(email=admin_email, role=UserRole.ADMIN))
            elif admin.role != UserRole.ADMIN:
                admin.role = UserRole.ADMIN

        await db.commit()


if __name__ == "__main__":
    asyncio.run(seed())

File: frontend/prisma/seed.ts
import { PrismaClient, AuthorBadge, AuthorType, ReviewStatus, UserRole } from '@prisma/client';

const prisma = new PrismaClient();

function normalize(value: string): string {
  return value.trim().toLocaleLowerCase();
}

async function main() {
  const adminEmail = process.env.ADMIN_EMAIL;
  if (!adminEmail) {
    throw new Error('ADMIN_EMAIL is required for seed');
  }

  const portugal = await prisma.country.upsert({
    where: { code: 'PT' },
    update: {},
    create: {
      code: 'PT',
      nameEn: 'Portugal',
      namePt: 'Portugal'
    }
  });

  const lisboa = await prisma.city.upsert({
    where: {
      id: 'seed-lisboa-city'
    },
    update: {},
    create: {
      id: 'seed-lisboa-city',
      countryId: portugal.id,
      name: 'Lisboa',
      normalizedName: normalize('Lisboa')
    }
  });

  const area = await prisma.area.upsert({
    where: {
      id: 'seed-alvalade-area'
    },
    update: {},
    create: {
      id: 'seed-alvalade-area',
      cityId: lisboa.id,
      name: 'Alvalade',
      normalizedName: normalize('Alvalade')
    }
  });

  const street = await prisma.street.upsert({
    where: { id: 'seed-rua-street' },
    update: {},
    create: {
      id: 'seed-rua-street',
      areaId: area.id,
      name: 'Rua de Exemplo',
      normalizedName: normalize('Rua de Exemplo')
    }
  });

  const segment = await prisma.streetSegment.upsert({
    where: { id: 'seed-segment-1' },
    update: {},
    create: {
      id: 'seed-segment-1',
      streetId: street.id,
      startNumber: 1,
      endNumber: 100
    }
  });

  const building = await prisma.building.upsert({
    where: {
      streetId_streetNumber: {
        streetId: street.id,
        streetNumber: 10
      }
    },
    update: {},
    create: {
      streetId: street.id,
      segmentId: segment.id,
      streetNumber: 10,
      lat: 38.757365,
      lng: -9.148911,
      buildingName: 'Prédio Exemplo'
    }
  });

  const adminUser = await prisma.user.upsert({
    where: { email: adminEmail.toLowerCase() },
    update: { role: UserRole.ADMIN },
    create: {
      email: adminEmail.toLowerCase(),
      role: UserRole.ADMIN,
      displayName: 'Admin'
    }
  });

  await prisma.review.upsert({
    where: { trackingCode: 'DEMO-APPROVED-1' },
    update: {},
    create: {
      buildingId: building.id,
      authorType: AuthorType.USER,
      authorUserId: adminUser.id,
      authorBadge: AuthorBadge.VERIFIED_ACCOUNT,
      status: ReviewStatus.APPROVED,
      trackingCode: 'DEMO-APPROVED-1',
      livedFromYear: 2022,
      livedToYear: 2024,
      livedDurationMonths: 24,
      peopleNoise: 4,
      animalNoise: 4,
      insulation: 3,
      pestIssues: 4,
      areaSafety: 4,
      neighbourhoodVibe: 5,
      outdoorSpaces: 4,
      parking: 3,
      buildingMaintenance: 3,
      constructionQuality: 3,
      overallScore: 3.7,
      overallScoreRounded: 3.7,
      comment: 'Good location and calm nights. Some maintenance delays.',
      piiFlagged: false,
      piiReasons: [],
      approvedAt: new Date(),
      lastModerationAction: 'APPROVE',
      moderatedByUserId: adminUser.id
    }
  });

  console.log('Seed complete');
}

main()
  .catch((error) => {
    console.error(error);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
